File: python/145_二叉树的后序遍历_medium.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# 迭代法
class Solution:
    def postorderTraversal(self, root: TreeNode) -> List[int]:
        stack = []
        ans = []
        pre = None
        while root or stack:
            while root:
                stack.append(root)
                root = root.left
            root = stack.pop()
            if not root.right or root.right == pre:
                ans.append(root.val)
                pre = root
                root = None
            else:
                stack.append(root)
                root = root.right
        return ans

# 递归
class Solution:
    def postorderTraversal(self, root: TreeNode) -> List[int]:
        def postOrder(root):
            # 注意递归的出口为空节点
            if not root:
                return
            postOrder(root.left)
            postOrder(root.right)
            ans.append(root.val)
        
        if not root:
            return []
        ans = []
        postOrder(root)
        return ans

File: cpp/145_二叉树的后序遍历_medium.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

// 迭代法
class Solution {
public:
    vector<int> postorderTraversal(TreeNode* root) {
        vector<TreeNode*> stack;
        vector<int> ans;
        TreeNode *pre;
        while(root || stack.size()){
            while(root){
                stack.push_back(root);
                root = root->left;
            }
            root = stack.back();
            stack.pop_back();
            if(!root->right || root->right == pre){
                ans.push_back(root->val);
                pre = root;
                root = nullptr;
            }
            else{
                stack.push_back(root);
                root = root->right;
            }
        }
        return ans; 
    }
};

// 递归法
class Solution {
public:
    vector<int> postorderTraversal(TreeNode* root) {
        vector<int> ans;
        postOrder(ans, root);
        return ans;
    }
    void postOrder(vector<int> &ans, TreeNode *root){
        if(!root)
            return;
        postOrder(ans, root->left);
        postOrder(ans, root->right);
        ans.push_back(root->val);
    }
};